src/consensus/merkle.h

```diff
@@ -90,4 +93,1361 @@ uint256 BlockWitnessMerkleRoot(const CBlock& block, bool* mutated = NULL);
+     * code, and therefore assignable to a MerkleNode, probably
+     * generating a memory access exception if the value is not
+     * between 0 and 7.
+     */
+    inline code_type GetCode() const
+      { return m_code; }
+
+    inline MerkleNode& SetCode(code_type code)
+    {
+        m_code = code;
+        return *this;
+    }
+
+    /*
+     * The getters and setters for the left and right MerkleLinks
+     * simply recalculate the code value using tables. The code values
+     * line up such that this could be done with arithmetic and
+     * shifts, but it is probably of similar efficiency.
+     */
+    inline MerkleLink GetLeft() const
+      { return m_left_from_code[m_code]; }
+
+    inline MerkleNode& SetLeft(MerkleLink left)
+    {
+        m_code = _get_code(left, m_right_from_code[m_code]);
+        return *this;
+    }
+
+    inline MerkleLink GetRight() const
+      { return m_right_from_code[m_code]; }
+
+    inline MerkleNode& SetRight(MerkleLink right)
+    {
+        m_code = _get_code(m_left_from_code[m_code], right);
+        return *this;
+    }
+
+    /* Equality */
+    inline bool operator==(MerkleNode other) const
+      { return (m_code == other.m_code); }
+    inline bool operator!=(MerkleNode other) const
+      { return !(*this == other); }
+
+    /* Relational */
+    inline bool operator<(MerkleNode other) const
+      { return (m_code < other.m_code); }
+    inline bool operator<=(MerkleNode other) const
+      { return !(other < *this); }
+    inline bool operator>=(MerkleNode other) const
+      { return !(*this < other); }
+    inline bool operator>(MerkleNode other) const
+      { return (other < *this); }
+
+    /* Needs access to m_{left,right}_from_code and _get_code() */
+    friend struct MerkleNodeReference;
+};
+
+/*
+ * Now we begin constructing the necessary infrastructure for
+ * supporting an STL-like container for packed 3-bit code
+ * representations of MerkleNode values. This is parallels the way
+ * that std::vector<bool> is specialized, with the added complication
+ * of a non-power-of-2 packed size.
+ */
+
+/*
+ * First we build a "reference" class which is able to address the
+ * location of a packed 3-bit value, and to read and write that value
+ * without affecting its neighbors.
+ *
+ * Then we will make use of this MerkleNode reference type to
+ * construct an STL-compatible iterator class (technically two, since
+ * the class const_iterator is not a const instance of the class
+ * iterator, for reasons).
+ */
+struct MerkleNodeReference
+{
+    /*
+     * Nodes are stored with a tightly packed 3-bit encoding, the
+     * code. This allows up to 8 node specifications to fit within 3
+     * bytes:
+     *
+     *    -- Node index
+     *   /
+     *   00011122 23334445 55666777
+     *    byte 0   byte 1   byte 2
+     *   76543210 76543210 76543210
+     *                            /
+     *                Bit Index --
```

Comment:
Is node at index 0 really located at bits 5-7 of byte 0? How does endianness of the system affect this? I assume it must not but it could be worthwhile to test this (I have a qemu debian big endian lying around, btw).

Reply:
That's how I access the bits and I do so in an endian-agnostic way. I access bytes directly, so this comment is more prescriptive than descriptive. But it would be nice to confirm that by running the test suite on a big endian machine.

As to why I access the bits this way, it is because it results in sane lexicographical ordering of proofs. Consider enumerating the leaves of a tree and interpreting a tree to be a representation of the list of leaf ordinals from left-to-right, so a Merkle tree with 4 leaves is the list [0 1 2 3]. A proof object is the same list with some elements removed, e.g. [1 2 3], [2], or [2 3]. With my purposeful layout of bits and selection of MerkleNode code values it turns out that if you lexicographically (byte-wise) sort such proofs of subsets of the same tree by their serialized representation, then the result is the same as if you compare the underlying list. So the proof for [1, 2, 3] (left-most leaf as SKIP) is lexicographically less than the proof for [2] (all but the 3rd element SKIP'd), which is less than the proof for [2, 3].

Is that useful? Maybe. But the choice of bit layout and node constants is otherwise arbitrary, so I might as well preserve some sensible properties to keep that option open.